src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.Cache.cs

```diff
@@ -24,71 +28,92 @@ public static int CacheSize
                 if (value < 0)
                     throw new ArgumentOutOfRangeException(nameof(value));
 
-                s_cacheSize = value;
-                if (s_livecode.Count > s_cacheSize)
+                lock (s_livecode)
                 {
-                    lock (s_livecode)
+                    s_cacheSize = value;  // not to allow other thread to change it while we use cache
+                    while (s_livecode.Count > s_cacheSize)
                     {
-                        while (s_livecode.Count > s_cacheSize)
-                            s_livecode.RemoveLast();
+                        CachedCodeEntry last = s_livecode_last;
+                        s_livecode.Remove(last.Key);
+
+                        // update linked list:
+                        s_livecode_last = last.Next;
+                        if (last.Next != null)
+                            last.Next.Previous = null;
+                        else  // last one removed
+                            s_livecode_first = null;
                     }
                 }
             }
         }
 
         /// <summary>
-        /// Find code cache based on options+pattern
+        ///  Find cache based on options+pattern+culture and optionally add new cache if not found
         /// </summary>
-        private static CachedCodeEntry LookupCachedAndUpdate(CachedCodeEntryKey key)
+        private CachedCodeEntry GetCachedCode(CachedCodeEntryKey key, bool isToAdd)
         {
+            if (s_cacheSize == 0)
+                return null;
             lock (s_livecode)
             {
-                for (LinkedListNode<CachedCodeEntry> current = s_livecode.First; current != null; current = current.Next)
+                // first look for it in the cache and move it to the head
+                var entry = LookupCachedAndPromote(key);
+                // it wasn't in the cache, so we'll add a new one
+                if (entry == null && isToAdd && s_cacheSize != 0)  // check cache size again in case it changed
                 {
-                    if (current.Value._key == key)
+                    entry = new CachedCodeEntry(key, capnames, capslist, _code, caps, capsize, _runnerref, _replref);
+                    s_livecode.Add(key, entry);
+                    // put in linked list:
+                    if (s_livecode_first != null)
+                    {
+                        s_livecode_first.Next = entry;
+                        entry.Previous = s_livecode_first;
+                    }
+                    s_livecode_first = entry;
```

Comment:
Similar comment about adding asserts throughout this to help validate the list's integrity.